### PR TITLE
[BEAM-3303] Add Go Windowing support

### PR DESCRIPTION
**Pull request by herohde**
* Added WindowInto and supporting types and handling. Made EventTime align with Beam
 * Added windowed_wordcount. Runs on direct + dataflow. Flink doesn't seem to support the urn yet.
 * Fixed Dataflow encoding issue and made it use model pipeline references for ParDo [BEAM-3448].

### CODE REVIEW

**Comment by herohde** [comment]
R: @wcn3 @lostluck 

Sorry for the massive PR.

File: sdks/go/pkg/beam/core/runtime/exec/combine.go [hunk 1]
```diff
@@ -93,51 +79,38 @@ func (n *Combine) ProcessElement(ctx context.Context, value FullValue, values ..
 		return fmt.Errorf("invalid status for combine %v: %v", n.UID, n.status)
 	}
 
-	if n.IsPerKey {
-		// For per-key combine, all processing can be done here. Note that
-		// we do not explicitly call merge, although it may be called implicitly
-		// when adding input.
+	// Note that we do not explicitly call merge, although it may
```
[the remark this answers is not in the document]
Comment (herohde) [reply sdks/go/pkg/beam/core/runtime/exec/combine.go]: It's referring to the lack of combiner lifting, where we precombine and then merge the results. We just add each input into one accumulator, for now.

File: sdks/go/pkg/beam/core/runtime/exec/combine.go [hunk 2]
```diff
@@ -93,51 +79,38 @@ func (n *Combine) ProcessElement(ctx context.Context, value FullValue, values ..
...
 
-		out, err := n.extract(ctx, a)
+		a, err = n.addInput(ctx, a, value.Elm, v.Elm, value.Timestamp, first)
 		if err != nil {
 			return n.fail(err)
 		}
-		return n.Out.ProcessElement(ctx, FullValue{Elm: value.Elm, Elm2: out, Timestamp: value.Timestamp})
+		first = false
 	}
+	stream.Close()
```
[the remark this answers is not in the document]
Comment (herohde) [reply sdks/go/pkg/beam/core/runtime/exec/combine.go]: Done

The defer cost is not a concern. I think the idea was to not hold the steam open while processing the continuation. We can tighten it up if it becomes an issue.

File: sdks/go/pkg/beam/core/runtime/exec/fn.go [hunk 3]
```diff
@@ -46,17 +48,22 @@ func Invoke(ctx context.Context, fn *funcx.Fn, opt *MainInput, extra ...interfac
 	if index, ok := fn.Context(); ok {
 		args[index] = ctx
 	}
+	if index, ok := fn.Window(); ok {
+		if len(ws) != 1 {
+			return nil, fmt.Errorf("DoFns that observe windows must be invoked with single window: %v", opt.Key.Windows)
```
[the remark this answers is not in the document]
Comment (herohde) [reply sdks/go/pkg/beam/core/runtime/exec/fn.go]: Each element may be in multiple windows, but are processing only once if the windows are not observed. Sliding windows is an example that produce multiple windows for an element.

File: sdks/go/pkg/beam/core/runtime/exec/pardo.go [hunk 4]
```diff
@@ -95,26 +96,56 @@ func (n *ParDo) ProcessElement(ctx context.Context, elm FullValue, values ...ReS
 	}
 
 	ctx = metrics.SetPTransformID(ctx, n.PID)
+	fn := n.Fn.ProcessElementFn()
 
-	val, err := n.invokeDataFn(ctx, elm.Timestamp, n.Fn.ProcessElementFn(), &MainInput{Key: elm, Values: values})
-	if err != nil {
-		return n.fail(err)
-	}
+	// If the function observes windows, we must pass invoke the it for each window. The expected fast path
```
[the remark this answers is not in the document]
Comment (herohde) [reply sdks/go/pkg/beam/core/runtime/exec/pardo.go]: Done

File: sdks/go/pkg/beam/core/runtime/exec/translate.go [hunk 5]
```diff
@@ -68,15 +69,19 @@ func UnmarshalPlan(desc *fnpb.ProcessBundleDescriptor) (*Plan, error) {
...
 					return nil, err
 				}
+				u.Coder = coder.NewW(c, wc)
 			} else {
 				u.Coder, err = b.coders.Coder(cid) // Expected to be windowed coder
 				if err != nil {
 					return nil, err
 				}
+				if !coder.IsW(u.Coder) {
+					return nil, fmt.Errorf("unwindowed coder %v on DataSource %v: %v", cid, id, u.Coder)
```
[the remark this answers is not in the document]
Comment (herohde) [reply sdks/go/pkg/beam/core/runtime/exec/translate.go]: Runner bug.

File: sdks/go/pkg/beam/core/runtime/exec/window.go [hunk 6]
```diff
@@ -0,0 +1,92 @@
...
+		var ret []typex.Window
+
+		period := mtime.FromDuration(wfn.Period)
+		lastStart := ts - (ts.Add(wfn.Size) % period)
+		for start := lastStart; start > ts.Subtract(wfn.Size); start -= period {
+			ret = append(ret, window.IntervalWindow{Start: start, End: start.Add(wfn.Size)})
+		}
+		return ret
+
+	default:
```
[the remark this answers is not in the document]
Comment (herohde) [reply sdks/go/pkg/beam/core/runtime/exec/window.go]: Opened BEAM-4152.

**Comment by herohde** [comment]
R: @robertwb Hope this PR makes Go a real SDK in your eyes :) PTAL

File: sdks/go/pkg/beam/core/graph/mtime/time.go [hunk 7]
```diff
@@ -0,0 +1,124 @@
...
+	"fmt"
+	"math"
+	"time"
+)
+
+const (
+	// MinTimestamp is the minimum value for any Beam timestamp. Often referred to
+	// as "-infinity". This value and MaxTimestamp are chosen so that their
+	// microseconds-since-epoch can be safely represented with an int64 and boundary
+	// values can be represented correctly with milli-second precision.
```
Comment (wcn3) [inline sdks/go/pkg/beam/core/graph/mtime/time.go]: millisecond

File: sdks/go/pkg/beam/core/runtime/exec/coder.go [hunk 8]
```diff
@@ -276,42 +277,135 @@ func (c *kvDecoder) Decode(r io.Reader) (FullValue, error) {
 
 }
 
-// TODO(herohde) 4/7/2017: actually handle windows.
+// WindowEncoder handles Window serialization to a byte stream. The encoder
+// can be reused, even if an error is encountered.
```
Comment (wcn3) [inline sdks/go/pkg/beam/core/runtime/exec/coder.go]: Any concurrency-safety requirements?

File: sdks/go/pkg/beam/core/runtime/exec/pardo.go [hunk 9]
```diff
@@ -95,26 +96,56 @@ func (n *ParDo) ProcessElement(ctx context.Context, elm FullValue, values ...ReS
 	}
 
 	ctx = metrics.SetPTransformID(ctx, n.PID)
+	fn := n.Fn.ProcessElementFn()
 
-	val, err := n.invokeDataFn(ctx, elm.Timestamp, n.Fn.ProcessElementFn(), &MainInput{Key: elm, Values: values})
-	if err != nil {
-		return n.fail(err)
-	}
+	// If the function observes windows, we must pass invoke it for each window. The expected fast path
```
Comment (wcn3) [inline sdks/go/pkg/beam/core/runtime/exec/pardo.go]: Grammar in first sentence. Perhaps, ", we must invoke it for each window?"

File: sdks/go/pkg/beam/core/runtime/graphx/coder.go [hunk 10]
```diff
@@ -75,15 +74,15 @@ type CoderUnmarshaller struct {
 	models map[string]*pb.Coder
 
 	coders  map[string]*coder.Coder
-	windows map[string]*window.Window
+	windows map[string]*coder.WindowCoder
```
Comment (wcn3) [inline sdks/go/pkg/beam/core/runtime/graphx/coder.go]: windowCoders?

File: sdks/go/pkg/beam/core/runtime/graphx/serialize.go [hunk 11]
```diff
@@ -942,28 +982,28 @@ func isCoGBKList(ref *CoderRef) ([]*CoderRef, bool) {
...
 		return &CoderRef{Type: GlobalWindowType}, nil
+	case coder.IntervalWindow:
+		return &CoderRef{Type: IntervalWindowType}, nil
 	default:
-		return nil, fmt.Errorf("bad window kind: %v", w.Kind())
+		return nil, fmt.Errorf("bad window kind: %v", w.Kind)
 	}
 }
 
 // decodeWindow receives the wire representation of a Beam coder, extracting
```
Comment (wcn3) [inline sdks/go/pkg/beam/core/runtime/graphx/serialize.go]: Fix name of function.

File: sdks/go/pkg/beam/runners/dataflow/translate.go [hunk 12]
```diff
@@ -454,50 +457,52 @@ func stepID(id int) string {
...
+	data, err := proto.Marshal(in)
+	if err != nil {
+		return "", err
+	}
+	return encodeString(data), nil
+}
+
+// encodeString is a custom encoding used in some cases by Dataflow.
+//
+// Uses a simple strategy of converting each byte to a single char,
```
Comment (wcn3) [inline sdks/go/pkg/beam/runners/dataflow/translate.go]: This is defined as 'percent encoding' or less accurately as 'URL encoding.' 
https://en.wikipedia.org/wiki/Percent-encoding

Pretty sure this is identical to url.PathEscape, perhaps use that?

File: sdks/go/pkg/beam/core/graph/coder/coder.go [hunk 13]
```diff
@@ -146,9 +145,9 @@ type Coder struct {
 	Kind Kind
 	T    typex.FullType
 
-	Components []*Coder       // WindowedValue, KV, CoGBK
-	Custom     *CustomCoder   // Custom
-	Window     *window.Window // WindowedValue
+	Components []*Coder     // WindowedValue, KV, CoGBK
+	Custom     *CustomCoder // Custom
+	Window     *WindowCoder // WindowedValue
```
[the remark this answers is not in the document]
Comment (herohde) [reply sdks/go/pkg/beam/core/graph/coder/coder.go]: No. The comments indicate which fields are populated for each kind as a poor man's algebraic data types.

File: sdks/go/pkg/beam/core/graph/mtime/time.go (hunk 7 above)
Comment (herohde) [reply sdks/go/pkg/beam/core/graph/mtime/time.go]: Done

File: sdks/go/pkg/beam/core/runtime/exec/coder.go (hunk 8 above)
Comment (herohde) [reply sdks/go/pkg/beam/core/runtime/exec/coder.go]: Added comment to all coding types that they are concurrency-safe.

File: sdks/go/pkg/beam/core/runtime/exec/pardo.go (hunk 9 above)
Comment (herohde) [reply sdks/go/pkg/beam/core/runtime/exec/pardo.go]: Done

File: sdks/go/pkg/beam/core/runtime/graphx/coder.go (hunk 10 above)
Comment (herohde) [reply sdks/go/pkg/beam/core/runtime/graphx/coder.go]: Done

File: sdks/go/pkg/beam/core/runtime/graphx/serialize.go (hunk 11 above)
Comment (herohde) [reply sdks/go/pkg/beam/core/runtime/graphx/serialize.go]: Done

File: sdks/go/pkg/beam/runners/dataflow/translate.go (hunk 12 above)
Comment (herohde) [reply sdks/go/pkg/beam/runners/dataflow/translate.go]: Seems you're right. Good catch.

**Comment by herohde** [comment]
Thanks!

**Comment by robertwb** [comment]
Thanks!

It looks like this structure only supports a finite set of builtin WindowFns, rather than letting the user define their own with a callable (WindowFn:WindowInto is like DoFn:ParDo), but that can be future work.

**Comment by herohde** [comment]
Thanks. Yes, the idea is add a custom windowFn later (similarly to how custom coders are done). My understanding is that custom WindowFns are not yet supported in the FnAPI, though.

**Pull request merged**